File: c++/0007. Valid Anagram.cpp
// time = O(n)
// space = O(n)

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.length() != t.length()) return false;
        unordered_map<char, int> freq;
        for (auto &c: s) {
            freq[c]++;
        }
        for (auto &c: t) {
            freq[c]--;
        }
        for (auto &pair: freq) {
            if (pair.second != 0) return false;
        }
        return true;
    }
};

File: py3/0007_valid_anagram.py
# time: O(n)
# space: O(n)
from collections import Counter
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        s_count = Counter(s)
        for c in t:
            if c not in s_count:
                return False
            s_count[c] -= 1
        return not any(i != 0 for i in s_count.values())